azurelinuxagent/ga/extension_telemetry.py

```diff
@@ -0,0 +1,385 @@
+
+    @staticmethod
+    def _get_extension_events_dir_with_handler_name(extension_log_dir):
+        """
+        Get the full path to events directory for all extension handlers that have one
+        :param extension_log_dir: Base log directory for all extensions
+        :return: A list of full paths of existing events directory for all handlers
+        """
+        extension_handler_with_event_dirs = []
+
+        for ext_handler_name in os.listdir(extension_log_dir):
+            # Check if its an Extension directory
+            if not os.path.isdir(os.path.join(extension_log_dir, ext_handler_name)) \
+                    or re.match(HANDLER_NAME_PATTERN, ext_handler_name) is None:
+                continue
+
+            # Check if EVENTS_DIRECTORY directory exists
+            extension_event_dir = os.path.join(extension_log_dir, ext_handler_name, EVENTS_DIRECTORY)
+            if os.path.exists(extension_event_dir):
+                extension_handler_with_event_dirs.append((ext_handler_name, extension_event_dir))
+
+        return extension_handler_with_event_dirs
+
+    def _capture_extension_events(self, handler_name, handler_event_dir_path, events_list):
+        """
+        Capture Extension events and add them to the events_list
+        :param handler_name: Complete Handler Name. Eg: Microsoft.CPlat.Core.RunCommandLinux
+        :param handler_event_dir_path: Full path. Eg: '/var/log/azure/Microsoft.CPlat.Core.RunCommandLinux/events'
+        :param events_list: List of captured extension events
+        """
+
+        convert_to_mb = lambda x: (1.0 * x)/(1000 * 1000)
+
+        # Filter out the files that do not follow the pre-defined EXTENSION_EVENT_FILE_NAME_REGEX
+        event_files = [event_file for event_file in os.listdir(handler_event_dir_path) if
+                       re.match(self.EXTENSION_EVENT_FILE_NAME_REGEX, event_file) is not None]
+        # Pick the latest files first, we'll discard older events if len(events) > MAX_EVENT_COUNT
+        event_files.sort(reverse=True)
+
+        captured_extension_events_count = 0
+        dropped_events_with_error_count = defaultdict(int)
+
+        for event_file in event_files:
+
+            event_file_path = os.path.join(handler_event_dir_path, event_file)
+            try:
+                logger.verbose("Processing event file: {0}", event_file_path)
+
+                # We only support EXTENSION_EVENT_FILE_MAX_SIZE=4Mb max file size
+                event_file_size = os.stat(event_file_path).st_size
+                if event_file_size > self.EXTENSION_EVENT_FILE_MAX_SIZE:
+                    msg = "Skipping file: {0} as its size is {1:.2f} Mb > Max size allowed {2:.1f} Mb".format(
+                            event_file_path, convert_to_mb(event_file_size),
+                            convert_to_mb(self.EXTENSION_EVENT_FILE_MAX_SIZE))
+                    logger.warn(msg)
+                    add_log_event(level=logger.LogLevel.WARNING, message=msg, forced=True)
```

Comment:
why add_log_event? (instead of add_event)

Reply:
add_event goes to the ExtensionEvents table and add_log_event goes to the GenericLogs table. I figured for the extension related stuff, it would make sense if all the data (including errors) are in 1 table so that its easier for debugging.

Reply:
But above you are using `add_event(op=WALAEventOperation.ExtensionTelemetryEventProcessing, message=msg, is_success=False)`

Reply:
The idea is basically that any error thrown by the agent while processing Extension Telemetry will be logged in ExtensionEvents table and any errors thrown due to the telemetry events (malformed, limit exceeds, etc) will be logged in the GenericLogs table for easier lookup